MapboxCoreNavigation/UpcomingRouteAlertInfo.swift

```diff
@@ -0,0 +1,84 @@
+import Foundation
+import CoreLocation
+import MapboxNavigationNative
+import MapboxDirections
+
+/**
+ `UpcomingRouteAlertInfo` encapsulates information about various incoming events. Common attributes like location, distance to the event, length and other is provided for each POI, while specific meta data is supplied via `alert` property.
+ */
+public struct UpcomingRouteAlertInfo {
+    public enum RouteAlertType {
+        case incident(Incident)
+        case tunnel(TunnelInfo)
+        case borderCrossing(BorderCrossingInfo)
+        case tollCollection(TollCollection)
+        case serviceArea(RestStop)
+    }
+    
+    /// Alert data with specific info. Contents depend on exact alert type.
+    public let alert: RouteAlertType
+    
+    /// Distance to route alert relative to start of the route, meters.
+    public let distance: CLLocationDistance
+    /**
+     Distance from current position to alert, meters.
+     
+     This value can be negative if it is a spanned alert and we are somewhere in the middle of it.
+     */
+    public let distanceToStart: CLLocationDistance
+    /**
+     Length of the alert info.
+     
+     This value will be non-null for composite route alerts */
+    public let length: CLLocationDistance?
+    
+    /// Coordinate of route alert beginning point
+    public let beginCoordinate: CLLocationCoordinate2D
+    /// Coordinate of route alert ending point
+    public let endCoordinate: CLLocationCoordinate2D
+    
+    /// Segment index in corresponding `Route.shape` where this alert begins.
+    public let beginSegmentIndex: UInt32
+    /// Segment index in corresponding `Route.shape` where this alert ends.
+    public let endSegmentIndex: UInt32
+    
+    init?(_ upcomingAlert: UpcomingRouteAlert) {
+        self.distance = upcomingAlert.alert.distance
+        self.distanceToStart = upcomingAlert.distanceToStart
+        self.length = upcomingAlert.alert.length?.doubleValue
+        self.beginCoordinate = upcomingAlert.alert.beginCoordinate
+        self.endCoordinate = upcomingAlert.alert.endCoordinate
+        self.beginSegmentIndex = upcomingAlert.alert.beginGeometryIndex
+        self.endSegmentIndex = upcomingAlert.alert.endGeometryIndex
+        
+        switch upcomingAlert.alert.type {
+        case .kIncident:
+            guard let incidentInfo = upcomingAlert.alert.incidentInfo else {
+                return nil
+            }
+            self.alert = .incident(Incident(incidentInfo))
+        case .kTunnelEntrance:
+            guard let tunnelInfo = upcomingAlert.alert.tunnelInfo else {
+                return nil
+            }
```

Comment:
Is this initializer failable because the alert may come back with no metadata? Or are these guard statements meant to guard against internal inconsistency? If the latter, perhaps these code paths should be `preconditionFailure`s.

Reply:
Hm... Originally I decided to make this one failable to cover cases when NN submits unknown alert type, so that SDK could just omit it. But it seems that such cases are impossible because NN version will always be in sync with SDK version, so having `default` case here might even become a potential point to forget add support for new alert types.
About optional alert info structures - these should not be nil if alert type is correct, so I think yes, that might be considered as an error.

Reply:
Ideally we’d add default unknown cases, but unfortunately, MapboxNavigationNative incorrectly declares all its structs as frozen.

Reply:
I thought that having a `default` case might create a room for an error once any new alert type is added to NN. If such happens now - compiler will alarm.

Reply:
Will it reliably produce an error if MapboxNavigationNative enables module stability?